feat(sheet): add documentation

Francois-Esquire opened the pull request:
## Description

This PR adds documentation for the `Sheet` component.

## Detail

The PR adds the `Sheet` component with 5 examples and documentation for usage and API.

## Checklist

- [x] :ok_hand: website updates are Garden Designer approved (add the designer as a reviewer)
- [ ] :black_nib: copy updates are approved (add the content strategist as a reviewer)
- [x] :link: considered opportunities for adding cross-reference URLs (grep for keywords)
- [x] :wheelchair: analyzed via [axe](https://www.deque.com/axe/) and evaluated using VoiceOver
- [x] :memo: tested in Chrome, Firefox, Safari, Edge, and IE11

Francois-Esquire commented on src/nav/components.yml:
```diff
@@ -48,6 +48,8 @@
           title: Toggle icon button
     - title: Chrome
       id: /components/chrome
+    - title: Sheet
+      id: /components/sheet
```
Depending on where the team thinks the Sheet component should be placed in the nav, we can move around accordingly.

jzempel replied:
Let's respect current alphabetical ordering.

jzempel replied:
btw, we can see here that "Color picker" and "Color swatch" are out of order – might be nice to sweep those up, too.

Francois-Esquire commented on src/pages/components/sheet.mdx:
```diff
@@ -0,0 +1,169 @@
+
+<PropSheet components={props.data.mdx.components} componentName="sheet" />
+
+### Sheet.Body
+
+The Sheet Body component is placed in the [`Sheet`](#sheet) component.
+This component requires accessible functionality to make the
+element focusable, if keyboard scrolling is needed.
+
+<Component components={props.data.mdx.components} componentName="sheet.body" />
+
+<PropSheet components={props.data.mdx.components} componentName="sheet.body" />
```
It seems the any component other than `Sheet` doesn't have a given `Extends HTMLAttributes<HTMLElement>` attributed. Is there something missing on my end on why this is missing?

Francois-Esquire replied:
Example:
<img width="789" alt="Screen Shot 2022-01-21 at 3 06 20 PM" src="https://user-images.githubusercontent.com/12474067/150593344-76a8f59f-c6e5-45ba-9ce4-d74a813619e1.png">

jzempel replied:
In its current form, you are responsible for fulfilling the JSDoc API for subcomponents. See https://github.com/zendeskgarden/website/blob/main/src/examples/accordions/components.ts (data) and https://github.com/zendeskgarden/website/blob/main/src/pages/components/accordion.mdx (usage) for details.

With the follow-on to https://github.com/zendeskgarden/react-components/pull/1273, I'm expecting to automate the data gathering for subcomponents, just as it is for top-level element components.

jzempel commented on src/examples/chrome/SheetSize.tsx:
```diff
@@ -0,0 +1,83 @@
+import { Button } from '@zendeskgarden/react-buttons';
+import { Fieldset, Field, Radio, Label } from '@zendeskgarden/react-forms';
+
+const [col4, col6, px380, px480] = ['4', '6', '380px', '480px'];
```
It feels like all the following codez would read simpler if the mapping could somehow be expressed up front as a constant. The behavior should be...

Col `size={4}` is Sheet `size="100%"`
Col `size={6}` is Sheet `size="100%"`
Sheet `size="380px"` is Col `size="auto"`
Sheet `size="480px"` is Col `size="auto"`

jzempel commented on src/examples/chrome/SheetSize.tsx:
```diff
@@ -0,0 +1,83 @@
+        </Col>
+
+        <Col size={size.endsWith('px') ? 9 : Number.parseInt(size, 10)}>
+          <div
+            style={{ width: '100%', height: '500px', display: 'flex', justifyContent: 'flex-end' }}
+          >
```
Per previous comments, prefer Garden grid layout over extra `div`

jzempel commented on src/examples/chrome/SheetSize.tsx:
```diff
@@ -0,0 +1,83 @@
+            {[col4, col6, px380, px480].map((sampleSize: string) => (
+              <Field key={`sizes-${sampleSize}`}>
+                <Radio
+                  name="default example"
```
`name` not needed for a non-submitted layout control

jzempel commented on src/pages/components/sheet.mdx:
```diff
@@ -0,0 +1,199 @@
+
+### Sheet.Close
+
+The Sheet Close button component is placed in the [`Sheet`](#sheet) component.
```
Is it worth identifying how it should be last for consistent tab order?

Francois-Esquire replied:
Good call on this.

jzempel commented on src/pages/components/sheet.mdx:
```diff
@@ -0,0 +1,199 @@
+
+### Sheet.Footer
+
+The Sheet Footer component is placed in the [`Sheet`](#sheet) component.
+The component can alternate to compact styling.
```
Please reference other component documentation for use of active voice. Also, something about this "...can alternate" feels a bit off grammatically.

Francois-Esquire replied:
Will update this and some of the text.

jzempel reviewed (changes requested):
Were the examples designed for responsive layout? The Garden website is meant to provide mobile responsive throughout. Please re-check all the examples. I'd recommend using `<Col sm>` for all of the 3-column layouts and then reference code under https://garden.zendesk.com/components/button-group#size for `StyledCol` and `mediaQuery` handling for stacked layouts. Also recommend working through the inline comments first (leaning into Garden layouts), as this should set things up for more consistent grid use.

It also looks like the API doc needs careful content review for passive voice.

james-sann commented:
> Were the examples designed for responsive layout? The Garden website is meant to provide mobile responsive throughout. Please re-check all the examples. I'd recommend using `<Col sm>` for all of the 3-column layouts and then reference code under https://garden.zendesk.com/components/button-group#size for `StyledCol` and `mediaQuery` handling for stacked layouts. Also recommend working through the inline comments first (leaning into Garden layouts), as this should set things up for more consistent grid use.
> 
> It also looks like the API doc needs careful content review for active vs passive voice.

@jzempel Are you asking the Garden designers to provide responsive design solutions for the component documentation page? If so, where can we find the breakpoints that might help guide @Francois-Esquire? @jon-kerwin were you aware of these? Happy to help. Just need the process shown to me!

jzempel commented:
> Are you asking the Garden designers to provide responsive design solutions for the component documentation page?

No. In the past, these have been worked out on-the-fly during implementation. However, if we feel that a process adaptation is necessary, breakpoint values can be found here: https://garden.zendesk.com/components/theme-object

diana-oum commented on src/pages/components/sheet.mdx:
```diff
@@ -0,0 +1,199 @@
+### Size
+
+Sheets default to 380px in width. You can customise the width based on your intended
+experience to any size using pixels or our grid system.
```
Sorry if this was discussed as I've been out of the loop but looking at the Figma file, "our grid system" is linked to https://garden.zendesk.com/components/grid

Francois-Esquire replied:
Good call, I'll update and link to Grid.

Francois-Esquire commented on src/examples/chrome/SheetDefault.tsx:
```diff
@@ -0,0 +1,80 @@
+        </Col>
+
+        <Col size={8}>
+          <div style={{ width: '100%', height: '500px', display: 'flex' }}>
+            {/* the div below represents main content sharing space with the Sheet in a container */}
+            <div style={{ flexGrow: 1 }} />
```
Sounds good, will remove the divs from the examples.

Francois-Esquire commented on src/examples/chrome/SheetFooters.tsx:
```diff
@@ -0,0 +1,66 @@
+
+      <Col size={4}>
+        <Sheet isOpen size="100%" style={surroundingBorderStyle}>
+          {/* div element below emulates Sheet.Body in its absence */}
+          <div style={{ flexGrow: 1 }} />
```
Good call, will update this.

Francois-Esquire commented on src/examples/chrome/SheetFooters.tsx:
```diff
@@ -0,0 +1,66 @@
+
+const surroundingBorder = '1px dotted gray';
+const surroundingBorderStyle = {
+  borderRight: surroundingBorder,
```
Yes good point, missed this originally, will update.

mtomcal replied:
Ill echo @jzempel on his other comment https://github.com/zendeskgarden/website/pull/337/files#r790861329 about replacing this object with a reusable styled component.

Francois-Esquire commented on src/examples/chrome/SheetFooters.tsx:
```diff
@@ -0,0 +1,66 @@
+          <div style={{ flexGrow: 1 }} />
+
+          <Sheet.Footer>
+            {/* margin-left was set to 0 due to column width wrapping the two buttons */}
+            <Sheet.FooterItem style={{ marginLeft: '0' }}>
```
Agreed, this was something that came to mind while working on the documentation.

Francois-Esquire commented on src/examples/chrome/SheetFooters.tsx:
```diff
@@ -0,0 +1,66 @@
+
+          <Sheet.Footer isCompact>
+            <Sheet.FooterItem>
+              <Anchor>Anchor</Anchor>
```
Agreed, it is off center, will update.

Francois-Esquire commented:
> > Were the examples designed for responsive layout? The Garden website is meant to provide mobile responsive throughout. Please re-check all the examples. I'd recommend using `<Col sm>` for all of the 3-column layouts and then reference code under https://garden.zendesk.com/components/button-group#size for `StyledCol` and `mediaQuery` handling for stacked layouts. Also recommend working through the inline comments first (leaning into Garden layouts), as this should set things up for more consistent grid use.
> > It also looks like the API doc needs careful content review for active vs passive voice.
> 
> @jzempel Are you asking the Garden designers to provide responsive design solutions for the component documentation page? If so, where can we find the breakpoints that might help guide @Francois-Esquire? @jon-kerwin were you aware of these? Happy to help. Just need the process shown to me!

Thanks @james-sann & @jzempel  for your help with thinking through the responsive design! I've updated the Sheet page with what we discussed offline and through the review. Please do let me know if I missed anything.

jzempel commented on src/pages/components/sheet.mdx:
```diff
@@ -0,0 +1,201 @@
+
+The `Sheet` component provides state and accessibility attributes to its children.
+
+<PropSheet components={props.data.mdx.components} componentName="sheet" />
```
It looks like there are some stray asterisks following prop descriptions that need to be cleaned-up as a TODO on the `react-components` codebase.

Francois-Esquire replied:
Raised [PR #1292](https://github.com/zendeskgarden/react-components/pull/1292) to take care of this.

james-sann reviewed (changes requested):
@Francois-Esquire Great work! Thanks for updating the responsive version.

Within the Placement section, we discussed making the container 90% on mobile viewports so there was a clear delineation of placing at the start or not. I couldn't see this reflected in the build. Just wondering if this was consciously addressed or not 🙌

jzempel commented on src/examples/chrome/SheetFooters.tsx:
```diff
@@ -0,0 +1,66 @@
+
+const Example = () => (
+  <Grid gutters="sm">
+    <Row style={{ height: '525px' }}>
```
@mtomcal we can't do shared components for examples, as they must successfully open/execute in the fixed Codesandbox environment.

Francois-Esquire commented on src/pages/components/sheet.mdx:
```diff
@@ -0,0 +1,201 @@
+
+<PropSheet components={props.data.mdx.components} componentName="sheet" />
+
+### Sheet.Body
```
Good catch, will update this.

mtomcal commented on src/pages/components/sheet.mdx:
```diff
@@ -0,0 +1,201 @@
+
+<Component components={props.data.mdx.subcomponents} componentName="sheet.body" />
+
+Nest `Sheet.Body` in the [`Sheet`](#sheet) component.
+`Sheet.Body` requires additional accessible functionality to make the
+element focusable for keyboard scrolling.
```
Should this ☝️ be recorded somewhere about documentation guidelines? TIL

Francois-Esquire replied:
@jzempel I removed the backticks. I'll reach out to @smritimadan here for more guidance.

@mtomcal you raise a good point here, worth recording down.

Francois-Esquire commented on src/examples/chrome/SheetDefault.tsx:
```diff
@@ -0,0 +1,98 @@
+  border: ${props => props.theme.borderWidths.sm} dashed;
+  border-color: ${props => getColor('neutralHue', 400, props.theme)};
+
+  ${props => mediaQuery('down', 'sm', props.theme)} {
+    & > aside {
+      box-sizing: content-box;
+      width: 100%;
+    }
+
+    /* stylelint-disable */
+    & > aside > div {
+      min-width: 100%;
+    }
+    /* stylelint-enable */
+  }
```
I don't think `box-sizing` is needed here, I will remove it.

Francois-Esquire commented:
> @Francois-Esquire Great work! Thanks for updating the responsive version.
> 
> Within the Placement section, we discussed making the container 90% on mobile viewports so there was a clear delineation of placing at the start or not. I couldn't see this reflected in the build. Just wondering if this was consciously addressed or not 🙌

@james-sann I missed this frankly speaking, more than happy to update! Before I do, what size do we want to set the sheet as the baseline? Should we follow the designs to the T here (ranged from 496px and up) or otherwise?

> Once the factor is bigger than 10, it feels like there are diminishing returns for leveraging `space.base`. For example, if the base were changed to 5, you probably don't want the height to change from 480 to 600px. Same for three – 480 down to 360px.

@jzempel agree, happy to update with a bit more direction here. This is probably related to the above, we could use a set pixel size here for all the examples (not just the headers/footers examples) since I noticed on IE11, the height was completely flattened here:

![Screen Shot 2022-02-07 at 5 28 19 PM](https://user-images.githubusercontent.com/12474067/152882938-3b196125-acaa-4eb9-90f6-cfffdf346683.png)

Francois-Esquire commented on src/examples/chrome/SheetPlacement.tsx:
```diff
@@ -0,0 +1,91 @@
+        <StyledSheet isOpen placement={placement}>
+          <Sheet.Header>
+            <Sheet.Title>Placement</Sheet.Title>
+            <Sheet.Description>From side to side</Sheet.Description>
```
Needed to deviate from the original content here due to axe testing and the need for unique `aria-labelledby` & `aria-describedby` values. Will need a content check here, and happy to update to what we want.

jzempel replied:
To be fair, the axe warnings here are a trailing concern. The leading concern is aria attribute management on the component itself. I would advocate for allowing the warnings to show here (and honoring the content design), as they are representative of an underlying component issue.

Francois-Esquire commented on src/examples/chrome/SheetSize.tsx:
```diff
@@ -0,0 +1,97 @@
+        <Col size={size.cols}>
+          <Sheet isOpen isAnimated={false} size={size.sheet}>
+            <Sheet.Header>
+              <Sheet.Title>Size</Sheet.Title>
+              <Sheet.Description>Sizing the Sheet</Sheet.Description>
```
The same update here (as above comment) due to axe raising this as an issue.

jzempel replied:
ditto comment reply.

jzempel commented on src/examples/chrome/SheetFooters.tsx:
```diff
@@ -0,0 +1,79 @@
+import { getColor, mediaQuery } from '@zendeskgarden/react-theming';
+
+const StyledCol = styled(Col)`
+  height: ${props => props.theme.space.base * 120}px;
+
+  ${props => mediaQuery('down', 'sm', props.theme)} {
+    height: ${props => props.theme.space.base * 80}px;
```
My point here was that we should just use:

```
height: 480px;

${props => mediaQuery('down', 'sm', props.theme)} {
  height: 320px;
```

jzempel commented on src/examples/chrome/SheetFooters.tsx:
```diff
@@ -0,0 +1,97 @@
+      <StyledSheet
+        isOpen
+        size="100%"
+        aria-describedby=""
+        aria-labelledby=""
```
Per conversation, please remove all empty `aria-describedby/labelledby`. They don't contribute to a clear understanding of expected usage and are indicators that the component itself should be improved to add valid attributes only when `Title` and/or `Description` exist.

Francois-Esquire replied:
This has been updated!

james-sann reviewed (changes requested):
@Francois-Esquire Just checked the RTL versions of the **Headers** and **Footers** examples, and they look a tad broken (Solid grey border-left is missing?)

Within the **Footers** section on RTL, the footer buttons are also stacking. They don't appear stacked on the LTR option.

<img width="754" alt="image" src="https://user-images.githubusercontent.com/85654216/153806249-dc92b3c9-91ba-4a9c-8e30-1faffd8a6871.png">

Francois-Esquire commented:
> @Francois-Esquire Just checked the RTL versions of the **Headers** and **Footers** examples, and they look a tad broken (Solid grey border-left is missing?)
> 
> Within the **Footers** section on RTL, the footer buttons are also stacking. They don't appear stacked on the LTR option.
> 
> <img alt="image" width="754" src="https://user-images.githubusercontent.com/85654216/153806249-dc92b3c9-91ba-4a9c-8e30-1faffd8a6871.png">

Thanks for pointing this out @james-sann ! I made some updates that addresses these issues. Let me know if anything else arises.

mtomcal commented on src/pages/components/sheet.mdx:
```diff
@@ -0,0 +1,200 @@
+
+<Component components={props.data.mdx.subcomponents} componentName="sheet.header" />
+
+Nest a Header within the [Sheet](#sheet) component. If neither [Title][#sheettitle] nor
+[Description][#sheetdescription] are used, the Header should be omitted.
```
```suggestion
Nest a Header within the [Sheet](#sheet) component. If neither [Title](#sheettitle) nor
[Description](#sheetdescription) are used, the Header should be omitted.
```
Found this:
<img width="754" alt="Screen Shot 2022-02-14 at 3 43 41 PM" src="https://user-images.githubusercontent.com/1975639/153959261-aac58894-f274-40e8-92d1-48e8c4006839.png">

Francois-Esquire replied:
Ahh good find! Thank you.

mtomcal reviewed (changes requested):
Quick fix before approval

james-sann reviewed (approved):
Looks great Mike! Approved!

diana-oum commented on src/pages/components/sheet.mdx:
```diff
@@ -0,0 +1,200 @@
+
+### Size
+
+Sheets default to 380px in width. You can customise the width based on your intended
```
@Francois-Esquire Just caught this now. "customise" --> "customize" 
We use American English spelling. It's mentioned in my [How to write for Garden guidelines](https://zendesk.atlassian.net/wiki/spaces/GARDEN/pages/5372510361/How+to+write+for+Garden) which got published after we designed this documentation but going forward I'll be really pushing these guidelines for both designers and engineers so we can all catch them earlier in the process. cc: @james-sann @jon-kerwin 

I was able to catch most of these things when in it was in the documentation phase but sometimes they still slip through...

Francois-Esquire replied:
Ahh I missed that as well! Good catch Diana, thank you!

diana-oum reviewed (approved):
"customise" needs to be "customize" but other than that, approved!
